iam_groups_authn/sync.py

```diff
@@ -48,142 +51,192 @@ async def groups_sync(iam_groups, sql_instances, credentials, private_ip=False):
+    user_service,
+    credentials,
+    ip_type,
+):
+    """
+    Sync the IAM members of a single group to a single Cloud SQL instance.
+    """
+    try:
+        database_version = await instance_tasks[1]
+        # verify that group role for database won't exceed character limit
+        verify_group_role_length(group, group_roles, database_version)
+        # add missing IAM group members to database
+        add_users_task = asyncio.create_task(
+            add_missing_db_users(
+                user_service,
+                group_task,
+                instance_tasks[0],
+                instance,
+                database_version,
             )
+        )
 
-            # grant group role to IAM users who are missing it on database
-            grant_role_task = asyncio.create_task(
-                grant_iam_group_role(
-                    role_service,
-                    role,
-                    users_with_roles_task,
-                    group_tasks[group],
-                    database_version,
-                )
-            )
-            results = await asyncio.gather(
-                revoke_role_task, grant_role_task, return_exceptions=True
-            )
-            # raise exception if found
-            for result in results:
-                if issubclass(type(result), Exception):
-                    raise result
+        # initialize database connection pool
+        if database_version.is_mysql():
+            db = init_mysql_connection_engine(instance, credentials, ip_type)
+            role_service = MysqlRoleService(db)
+        else:
+            db = init_postgres_connection_engine(instance, credentials, ip_type)
+            role_service = PostgresRoleService(db)
+        logging.debug(
+            "[%s][%s] Initialized a %s connection pool."
+            % (instance, group, database_version.value)
+        )
 
-            # log sync info
-            revoked_users, granted_users = results
-            logging.info(
-                "[%s][%s] Sync successful: %s users were revoked group role, %s users were granted group role."
-                % (instance, group, len(revoked_users), len(granted_users))
-            )
-            logging.debug(
-                "[%s][%s] Users revoked role: %s." % (instance, group, revoked_users)
-            )
-            logging.debug(
-                "[%s][%s] Users granted role: %s." % (instance, group, granted_users)
+        # verify role for IAM group exists on database, create if does not exist
+        role = group_roles.get(group, mysql_username(group))
+        verify_role_task = asyncio.create_task(role_service.create_group_role(role))
+
+        # get database users who have group role
+        users_with_roles_task = asyncio.create_task(
+            get_users_with_roles(role_service, role)
+        )
+
+        # await dependent tasks
+        results = await asyncio.gather(
+            add_users_task, verify_role_task, return_exceptions=True
```

Comment:
Would it be better to not use return_execptions here and just let it throw whatever error happens first? 

> If return_exceptions is False (default), the first raised exception is immediately propagated to the task that awaits on gather()